org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/utils/XMLGenerator.java

```diff
@@ -134,23 +134,28 @@ public String generate(Collection<CMAttributeDeclaration> attributes, String tag
 
 	private int generate(Collection<CMAttributeDeclaration> attributes, int level, int snippetIndex, XMLBuilder xml, String tagName) {
 		int attributeIndex = 0;
-		int attributesSize = attributes.size();
-		for (CMAttributeDeclaration attributeDeclaration : attributes) {
-			if (attributeDeclaration.isRequired()) {
-				String value = attributeDeclaration.getDefaultValue();
-				if (canSupportSnippets) {
-					snippetIndex++;
-					value = ("$" + snippetIndex);
-				}
-				if(attributesSize != 1) {
-					xml.addAttributes(attributeDeclaration.getName(), value, attributeIndex, level, tagName);
-				}
-				else {
-					xml.addSingleAttribute(attributeDeclaration.getName(), value);
-				}
-				
-				attributeIndex++;
+		
+		ArrayList<CMAttributeDeclaration> requiredAttributes = new ArrayList<CMAttributeDeclaration>();
+		for (CMAttributeDeclaration att: attributes) {
+			if(att.isRequired()) {
+				requiredAttributes.add(att);
+			}
+		}
+		int attributesSize = requiredAttributes.size();

```

Comment:
and if there's no required attributes? no indentation?

Reply:
Yeah Im not sure why, but thats what angelo had. @angelozerr whats the reason for only using attribute.isRequired() attributes?

Reply:
If I understand your question, is why we generate only required attributes? If it that it's because when you apply completion, it generates only required attributes and not the whole attributes. Not sure that I have understood the question.

Reply:
XSD example:
``` 
<xs:attribute name="lang" type="xs:string" use="required"/>
```

See `use="required"` above.

Is that what isRequired() refers to?

And if we had an element that was optional:

```
<xs:attribute name="lang" type="xs:string" use="optional"/>
```

would isRequired() == false ?

Why is isRequired() necessary to be checked to consider the attribute as a completion item, are there too many "optional" attributes?